---
title: "dada2_dwca_test"
author: "PascalH"
date: "2023-06-02"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load packages

```{r}
#if (!requireNamespace("BiocManager", quietly = True))
#  install.packages("BiocManager")
#BiocManager::install("dada2", version = "3.14")

require(dada2)
```

## Set the path to the data

```{r}
path <- "../data/sequences" # change to the directory containing the fastq files after unzipping.
list.files(path)
```

## Make lists of file  names

```{r}
# Forward and reverse fastq filenames have format: SAMPLENAME_R1_001.fastq and SAMPLENAME_R2_001.fastq
fnFs <- sort(list.files(path, pattern="_1.fastq", full.names = TRUE))
fnRs <- sort(list.files(path, pattern="_2.fastq", full.names = TRUE))
# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)
```

## Inspect read quality profiles

```{r}
plotQualityProfile(fnFs[1:2]) # forward reads
plotQualityProfile(fnRs[1:2]) # reverse reads
```

## Filter and trim

```{r}
# Place filtered files in filtered/ subdirectory
filtFs <- file.path(path, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(path, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))
names(filtFs) <- sample.names
names(filtRs) <- sample.names

out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(240,160),
              maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=8) # On Windows set multithread=FALSE, uses all cores if set to TRUE, therefore an integer should be specified
head(out)
```

## Error rates

```{r}
errF <- learnErrors(filtFs, multithread=8)

errR <- learnErrors(filtRs, multithread=8)

plotErrors(errF, nominalQ=TRUE)
```

## Sample inference

```{r}
dadaFs <- dada(filtFs, err=errF, multithread=8)

dadaRs <- dada(filtRs, err=errR, multithread=8)

dadaFs[[1]]
```

## Merge read pairs

```{r}
mergers <- mergePairs(dadaFs, filtFs, dadaRs, filtRs, verbose=TRUE)
# Inspect the merger data.frame from the first sample
head(mergers[[1]])
```

## Construct sequence table

```{r}
seqtab <- makeSequenceTable(mergers)
dim(seqtab)

# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab)))
```

## Remove chimeras

```{r}
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)
sum(seqtab.nochim)/sum(seqtab)
```

## Track reads through the pipeline

```{r}
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
# If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)
colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
head(track)
```

## Assign taxonomy

```{r}
# Analysis run with only the first 150k lines of the pr2 database to avoid memory issues
taxa <- assignTaxonomy(seqtab.nochim, "../data/pr2_subset.dada2.fa.gz", taxLevels = c("Domain","Supergroup","Division","Subdivision","Class","Order","Family","Genus","Species"), multithread=8)
```

```{r}
taxa.print <- taxa # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)
```

## Occurrence and DNA extension table

```{r}
source("dwca_test.R")
output = dwca_test("../data/metadata_test.csv", seqtab.nochim, taxa, "../results")
```